### PR TITLE
Feature/sortable

### PR DESCRIPTION
**Pull request by jordanmaguire**
FG now supports sorting on the top level attributes. Will add in support for nested stuff a bit l8r.

This should save me a tonne of time working on Healthforce.

### CODE REVIEW

File: lib/frontier/factory_girl_support/declaration.rb [hunk 1]
```diff
@@ -7,8 +7,8 @@ def initialize(verb, factory_object)
     @factory_object = factory_object
   end
 
-  def to_s
-    "FactoryGirl.#{verb}(:#{class_name})"
+  def to_s(factory_options={})
+    "FactoryGirl.#{verb}(#{factory_body(factory_options)})"
```
Comment (BenR1312) [inline lib/frontier/factory_girl_support/declaration.rb]: Is this for generating premade factories? i dont really undestand what this is at all?

**Comment by BenR1312** [comment]
This is super cool man.
When there is some time do you reckon you could maybe do a tech talk about how youve set up the generator and explain some of the finer details of how its all pieced together?
Becuase it consistently looks like cool magic to me

File: lib/frontier/factory_girl_support/declaration.rb (hunk 1 above)
Comment (jordanmaguire) [inline lib/frontier/factory_girl_support/declaration.rb]: > Is this for generating premade factories? i dont really undestand what this is at all?

Check the specs :) It's generating a string that when injected inside a template will be generated as ruby code. That ruby code will evaluate to be a call to FactoryGirl

File: lib/frontier/feature_spec.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class Frontier::FeatureSpec
+end
+
+require_relative "./feature_spec/order_expectation_method"
+require_relative "./feature_spec/sort_scenario"
+require_relative "./feature_spec/visit_index_method"
+
```
Comment (pzi) [inline lib/frontier/feature_spec.rb]: oh hai ws

**Comment by pzi** [comment]
Robots, you gotta love'em.

magic

File: lib/frontier/controller_action/index_action.rb [hunk 3]
```diff
@@ -6,14 +6,37 @@ def to_s
     raw = <<-STRING
 def index
   #{Frontier::Authorization::Assertion.new(model_configuration, :index).to_s}
-  #{model_configuration.as_ivar_collection} = #{scopable_object}.page(params[:page])
+#{Frontier::RubyRenderer.new(object_scope).render(1)}
```
Comment (kevin-jackson) [inline lib/frontier/controller_action/index_action.rb]: Indenting

File: lib/frontier/controller_action/index_action.rb [hunk 4]
```diff
@@ -6,14 +6,37 @@ def to_s
...
+
+  def object_scope
+    if sortable?
+      ransack_query = "@ransack_query = #{model_configuration.as_constant}.ransack(params[:q])"
+      merged_query = "#{model_configuration.as_ivar_collection} = #{scopable_object}.merge(@ransack_query.result)"
+
+      # The leading whitespace will allow the second line of the query to line up with the first
+      leading_whitespace = Array.new(merged_query.split(".").first.length, " ").join("")
+      pagination  = "#{leading_whitespace}.page(params[:page])"
+
```
Comment (kevin-jackson) [inline lib/frontier/controller_action/index_action.rb]: Should there be a new line here?

File: lib/frontier/feature_spec/order_expectation_method.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  def to_s
+    raw = <<STRING
+def #{method_name}(first, second)
+  within(first_row)  { expect(page).to have_content(first.#{primary_attribute_name}) }
+  within(second_row) { expect(page).to have_content(second.#{primary_attribute_name}) }
+end
+STRING
+    raw.rstrip
```
Comment (colynprater) [inline lib/frontier/feature_spec/order_expectation_method.rb]: ooc what is goin on up in here?

File: spec/lib/frontier/factory_girl_support/declaration_spec.rb [hunk 6]
```diff
@@ -3,32 +3,55 @@
...
       end
     end
 
-    context "with a string" do
+    describe "providing additional_options" do
       let(:factory_object) { "yolotronix" }
-      it { should eq("FactoryGirl.build(:yolotronix)") }
+
+      context "with no options" do
+
```
Comment (colynprater) [inline spec/lib/frontier/factory_girl_support/declaration_spec.rb]: pending?

**Comment by colynprater** [comment]
some buckwild code :+1:

File: lib/frontier/controller_action/index_action.rb (hunk 3 above)
Comment (jordanmaguire) [inline lib/frontier/controller_action/index_action.rb]: it's intentional. The render method will format all lines at a given line. See the specs for more info

Comment (jordanmaguire) [inline lib/frontier/controller_action/index_action.rb]: See [this blog post](http://blog.jayfields.com/2006/12/ruby-multiline-strings-here-doc-or.html) for more info about heredoc, which is what line 6 is initiating and line 11 is ending.

File: lib/frontier/feature_spec/order_expectation_method.rb (hunk 5 above)
Comment (jordanmaguire) [inline lib/frontier/feature_spec/order_expectation_method.rb]: the way heredoc works...http://blog.jayfields.com/2006/12/ruby-multiline-strings-here-doc-or.html

There is a newline included between `end` on line 14 and `STRING` on line 15.

To get around that I could write `endSTRING` but that is super hard to read and text editors can't parse it. So instead I strip the newline off the right hand side of the string :)

File: spec/lib/frontier/factory_girl_support/declaration_spec.rb (hunk 6 above)
Comment (jordanmaguire) [inline spec/lib/frontier/factory_girl_support/declaration_spec.rb]: wow - totally missed it. Good catch!

**Pull request merged**